src/api_manager/rewrite_rule.cc

```diff
@@ -90,77 +135,153 @@ RewriteRule::RewriteRule(std::string regex, std::string replacement,
-        env_->LogError(rewrite_log.str());
-      }
-      return false;
-    }
+  if (pcre_exec_ret < 0) {  // Something bad happened..
+    if (debug_mode) {
+      std::string msg;
 
-    if (debug_mode_) {
-      rewrite_log << kEspRewriteTitle << ": regex=" << regex_pattern_
-                  << std::endl;
-      rewrite_log << kEspRewriteTitle << ": request uri=" << uri << std::endl;
-      for (size_t i = 0; i < cm.size(); i++) {
-        rewrite_log << kEspRewriteTitle << ": $" << std::to_string(i) << ": "
-                    << std::string(cm[i]) << std::endl;
+      switch (pcre_exec_ret) {
+        case PCRE_ERROR_NOMATCH:
+          msg.assign("String did not match the pattern");
+          break;
+        case PCRE_ERROR_NULL:
+          msg.assign("Something was null");
+          break;
+        case PCRE_ERROR_BADOPTION:
+          msg.assign("A bad option was passed");
+          break;
+        case PCRE_ERROR_BADMAGIC:
+          msg.assign("Magic number bad (compiled re corrupt?)");
+          break;
+        case PCRE_ERROR_UNKNOWN_NODE:
+          msg.assign("Something kooky in the compiled re");
+          break;
+        case PCRE_ERROR_NOMEMORY:
+          msg.assign("Ran out of memory");
+          break;
+        default:
+          msg.assign("Unknown error");
+          break;
       }
 
-      rewrite_log << kEspRewriteTitle << ": replacement: " << replacement_
+      rewrite_log << kEspRewriteTitle << ": matching rule: " << regex_pattern_
+                  << ", request uri: " << uri << ", error: " << msg
                   << std::endl;
+      env_->LogInfo(rewrite_log.str());
     }
 
-    std::stringstream ss;
-    for (auto it = replacement_parts_.begin(); it != replacement_parts_.end();
-         ++it) {
-      switch (it->type) {
-        case ReplacementPartType::TEXT:
-          ss << it->text;
-          break;
-        case ReplacementPartType::REPLACEMENT:
-          if (it->index >= 0 && it->index < (int)cm.size()) {
-            ss << cm[it->index];
-          }
-          break;
-      }
+    restore_regex_malloc_free();
+    return false;
+  }
+
+  if (debug_mode) {
+    rewrite_log << kEspRewriteTitle << ": matching rule: " << regex_pattern_
+                << " " << this->replacement_ << std::endl;
+  }
+
+  // At this point, rc contains the number of substring matches found...
+  if (pcre_exec_ret == 0) {
+    rewrite_log << kEspRewriteTitle
+                << ": too many substrings were found to fit in sub_str_vec!"
+                << std::endl;
+    // Set rc to the max number of substring matches possible.
+    pcre_exec_ret = kMaxRegexMathCount / 3;
+  }
+
+  std::vector<std::string> match;
+
+  const char *psub_str_match_str;
+  for (int j = 0; j < pcre_exec_ret; j++) {
+    pcre_get_substring(uri.c_str(), sub_str_vec, pcre_exec_ret, j,
+                       &(psub_str_match_str));
+    match.push_back(psub_str_match_str);
+  }
+  pcre_free_substring(psub_str_match_str);
```

Comment:
instead of making a copy,  can we do this in line 266

Reply:
Applied